Implement reference test support for rendering at different zoom levels (issue 14505)

quaoaris opened the pull request:
Add tests to check that the rendering is correct at different zoom levels issue: #14505

This PR adds a scale factor in test_manifest.json. The scale factor is set by default to 1.

Setting to a value >1 results in a larger image for the browsertest.

Snuffleupagus commented on test/add_test.js:
```diff
@@ -65,6 +65,7 @@ calculateMD5(file, (err, md5) => {
     `       "file": "pdfs/${pdf}.pdf",\n` +
     `       "md5": "${md5}",\n` +
     '       "rounds": 1,\n' +
+    '       "scale": 1,\n' +
```
Please remove this line, since this parameter should be optional and consequently only needs to be set then *not* equal to one (which isn't going to be that common).

By doing this you'll unnecessarily inflate the size of the `test_manifest.json` file, and I'd view this new parameter similar to the exiting (and optional) `firstPage`/`lastPage` parameters that we only set when actually needed.

Snuffleupagus commented on test/driver.js:
```diff
@@ -431,6 +431,9 @@ class Driver {
       task.stats = { times: [] };
       task.enableXfa = task.enableXfa === true;
 
+      // just in case scale is not set, setting to default=1
```
In this case, the comment doesn't really add all that much value (compared to simply reading the code); hence I'd suggest removing this one.

(For future reference, this type of comments should be a *complete* sentence. That means beginning with an upper-case letter and ending with a period.)

Snuffleupagus commented on test/driver.js:
```diff
@@ -431,6 +431,9 @@ class Driver {
       task.enableXfa = task.enableXfa === true;
 
+      // just in case scale is not set, setting to default=1
+      task.scale = task.scale || 1;
```
I believe that `task.scale ||= 1;` ought to work, given the browsers that we run tests in.

Snuffleupagus commented on test/driver.js:
```diff
@@ -601,7 +604,7 @@ class Driver {
           page => {
             const viewport = page.getViewport({
-              scale: PixelsPerInch.PDF_TO_CSS_UNITS,
+              scale: PixelsPerInch.PDF_TO_CSS_UNITS * task.scale,
```
Nit: To be consistent with e.g. the viewer usage, let's re-order this as `task.scale * PixelsPerInch.PDF_TO_CSS_UNITS` instead.

Snuffleupagus reviewed (changes requested):
Besides the inline comments, note that as-is this patch doesn't really "do anything" on its own.

Please actually add one (or more) "new" test-cases that actually use the `scale` parameter (with a non-default value) in the browser-tests; this was also mentioned in e.g. https://github.com/mozilla/pdf.js/issues/14505#issuecomment-1023470637.

Snuffleupagus commented:
Please revert all re-formatting of the `test_manifest.json` file, since those changes make it very difficult to actually review this and it's also completely unrelated to the changes you're making here.

quaoaris committed:
Implement reference test support for rendering at different zoom levels

Add tests to check that the rendering is correct at different zoom levels mozilla#14505

An optional "scale" factor in test_manifest.json scales the resulting image (up/down) in browsertest.

Snuffleupagus commented:
@calixteman What sort of additional test-cases (and how many) did you have in mind in https://github.com/mozilla/pdf.js/issues/14505#issuecomment-1023470637?

timvandermeij commented on test/test_manifest.json:
```diff
@@ -5954,6 +5954,15 @@
        "lastPage": 1,
        "type": "eq"
     },
+	{
```
This entry is misaligned compared to the other entries.

timvandermeij reviewed (commented):
I agree with the question above: while this looks good in general, is the intention just to have a test case to cover rendering with a different scale or did you intend to have more tests here, for example with different zoom levels?

Moreover, the commit message should start with a verb, i.e., something like `Implement reference test support for rendering at different zoom levels` should be more descriptive.

calixteman commented:
And I planned to add a test for #14521.

Snuffleupagus commented on test/test_manifest.json:
```diff
@@ -5954,6 +5954,15 @@
        "lastPage": 1,
        "type": "eq"
     },
+    {
+	   "id": "issue6127_scale4",
+	   "file": "pdfs/issue6127.pdf",
+	   "md5": "65ef9572bddef5fddfaf7172f10a0df2",
+	   "rounds": 1,
+	   "lastPage": 1,
+	   "scale": 4,
+	   "type": "eq"
+    },
```
https://github.com/mozilla/pdf.js/pull/14534#discussion_r800173739 unfortunately still applies here, since you've mixed tabs and spaces.

Please note that we *only* use spaces for indentation everywhere in the code-base. It might help to setup the EditorConfig integration for your code-editor to help with formatting, see e.g. https://editorconfig.org/#pre-installed or https://editorconfig.org/#download

---

> Setting to a value >1 results in a larger image for the browsertest.

*Just a FYI:* It'll also be possible to set the `scale` to a value *below* `1`, since one here simply means 100% (e.g. in the viewer).

Snuffleupagus commented:
Also, it seems that the *commit message* now contains some unnecessary lines that can/should be removed:
 - `formatting test_manifest.json` looks completely unrelated, and should be removed.
 - `eq test at different zoom levels (issue #14505 )` which was the old title of the commit, but can be removed now.

Snuffleupagus commented:
> I still have problems using Githubdesktop

Personally I'd *very strongly* suggest using the Git command line directly rather that going through some kind of UI since that should give you much greater control over things.
Interactive rebasing with `git rebase -i`, as mentioned e.g. in https://github.com/mozilla/pdf.js/wiki/Squashing-Commits, is usually the best/easiest way to fixup a bad commit series (and it's useful for other things as well).

timvandermeij commented:
Let's close this PR for now as it is over a year old now and outdated. Thanks.
